### ISSUE TITLE
[FEATURE] Add Data Packing Support for WAN2.2 Training

### ISSUE THREAD

**Issue by VEGA1215** [baidu-baige/LoongForge#57]
## Summary

This feature adds packing support for WAN2.2 training. It enables multiple variable-length video/text samples to be concatenated into one packed training sequence and uses THD-format packed attention metadata to preserve sample boundaries during self-attention, cross-attention, loss computation, and context-parallel execution.

## Motivation

WAN2.2 training data can contain samples with different latent video lengths, resolutions, and text lengths. Without packing, each sample must be padded to the maximum configured sequence length, which wastes compute on padding tokens and reduces effective throughput.

Packing addresses this by:

- Supporting variable-length dataset training for WAN2.2.
- Reducing unnecessary computation on padding tokens.
- Improving GPU utilization by filling each training sequence with real tokens from multiple samples.
- Keeping per-sample attention boundaries so packed samples do not attend to each other incorrectly.
- Preserving compatibility with context parallelism, including Ring, Ulysses, and hybrid Ring+Ulysses modes.

## Results

Numerical consistency was validated by comparing packing-on and packing-off training logs for the first 100 iterations. The table reports LM loss absolute error and percentage error, where the percentage error is computed relative to the packing-off LM loss.

| Scenario | Avg Abs LM Loss Error | Max Abs LM Loss Error | Avg LM Loss Error (%) | Max LM Loss Error (%) |
|---|---:|---:|---:|---:|
| No CP | `1.115e-05` | `5.910e-05` | `0.004849%` | `0.024159%` |
| Pure Ring CP | `1.107e-05` | `4.280e-05` | `0.004635%` | `0.018719%` |
| Pure Ulysses CP | `1.071e-05` | `6.830e-05` | `0.004350%` | `0.030264%` |
| Hybrid Ring + Ulysses CP | `9.964e-06` | `4.190e-05` | `0.004061%` | `0.018054%` |

Across all four configurations, the average LM loss error stays around `1e-5`, and the maximum percentage error remains below `0.031%`, indicating that packing-on and packing-off training are numerically consistent.

The remaining numerical difference is expected because packing-on uses THD packed attention while packing-off uses the dense SBHD path, which can dispatch different TransformerEngine attention kernels and produce minor BF16/FP16 accumulation differences. Detailed precision data is available in the packing-on and packing-off log files.

## Implementation

The WAN2.2 packing path introduces a packed forward flow that accepts `PackedSeqParams` and per-sample `grid_sizes`. Packed samples are represented in THD format, with cumulative sequence length metadata used to preserve sample boundaries.

The implementation includes:

- Packed dataset loading and collation for variable-length video samples.
- Construction of self-attention and cross-attention `PackedSeqParams` from packed batch metadata.
- Packed WAN model forward dispatch when `grid_sizes` is present.
- Per-sample RoPE construction for packed video latents.
- Per-sample timestep modulation inside packed transformer layers.
- Packed self-attention and cross-attention with THD metadata.
- Context-parallel support for packed sequences, including external THD all-to-all handling for Ulysses paths.
- Loss masking for packed samples so padding tokens do not contribute to the loss.

## Key Components

| File | Change |
|---|---|
| `loongforge/data/video/packed_dataset.py` | Adds packed video dataset support, groups variable-length samples, builds packed tensors, and emits `grid_sizes`, sequence lengths, loss masks, and packed batch metadata. |
| `loongforge/train/diffusion/pretrain_wan.py` | Detects WAN packing mode, builds `PackedSeqParams`, moves packed metadata to GPU, passes `packed_seq_params` and `grid_sizes` into the model, and applies packed loss masking. |
| `loongforge/models/diffusion/wan/wan_model.py` | Adds packed forward dispatch, builds per-sample packed RoPE frequencies, handles CP-aware packed sequence splitting, and keeps global packed sequence metadata for TE CP attention. |
| `loongforge/models/diffusion/wan/wan_layer.py` | Adds packed transformer-layer execution with per-sample modulation and packed self-attention/cross-attention metadata. |
| `loongforge/models/diffusion/wan/wan_attention.py` | Adds THD packed attention support and external THD all-to-all communication for Ulysses/hybrid CP modes. |
| `loongforge/train/validators.py` | Enables variable sequence lengths for packing and keeps the CP communication topology aligned between packing-on and packing-off runs. |

## Configuration

Enable WAN2.2 packing by adding the packing flag and an optional packing buffer size to the training arguments:

```bash
--packing-sft-data
--packing-buffer-size
```

`--packing-buffer-size` controls how many samples are loaded into the packing buffer before the packing algorithm groups them into packed sequences. A larger buffer can improve packing efficiency, while a smaller buffer can reduce host-side memory usage. When packing is enabled, `micro_batch_size` is automatically forced to `1` by the argument validator.

## Limitations

- Packing currently requires `micro_batch_size=1`; the validator forces this setting when packing is enabled.
- Packing-on and packing-off may not be bitwise identical because THD packed attention and SBHD dense attention can use different attention kernels.
- Small grad norm differences can be larger than LM loss differences because gradient norm aggregates squared gradient values across all parameters.
- The dataset must provide the packed metadata required by WAN training, including per-sample grid sizes and sequence length information.
